Add test for push/pop operators + Fix a slew of PC related issues

dezuman opened the pull request.

dezuman committed:
Clean up macro code, purge some obsolete patters

dezuman committed:
Fix for BL instruction using wrong addresses unless 4-aligned

dezuman committed:
Add test for BL op to test function with different code alignments

dezuman committed:
Fix Push and Pop behavior

dezuman committed:
Fix code obsoleted by rebase

dezuman committed:
Unignore helloworld test

dezuman committed:
WIP fix for branch alignment issue

dezuman committed:
Remove (apparently?) incorrect 4-alignment in jump address base calc

dezuman committed:
Update tests

dezuman committed:
Add memory checking to testing macros

dezuman committed:
Do rustfmt

dezuman committed:
Add test for PUSH/POP operators

dezuman committed:
Fix issue to a LDR op variant caused by PC changes 

Apparently that particular instruction actually needed PC aligned by 4, but that was to be fair specified explicitly in doc.

Earlz commented on src/narmvm.rs:
```diff
@@ -105,8 +105,10 @@ impl NarmVM{
                 let lr = LongRegister{register: 14};
-                self.set_reg(&lr, self.virtual_pc | (self.pc & 1)); //or with bottom bit of current pc to copy interworking mode
+                
+                self.set_reg(&lr, self.virtual_pc | 1); //or with bottom bit of current pc to copy interworking mode
```
self.pc & 1 is the more proper way to do this, but this is fine since we aren't planning to go beyond thumb mode anytime soon

dezuman replied:
Fixed it

Earlz commented on src/narmvm.rs:
```diff
@@ -970,10 +972,12 @@ impl NarmVM{
-            msg.push_str(&format!("[{:#010x}] -- {} -- {:#06x}\n",
+            msg.push_str(&format!("[{:#08x}] -- [{:#04}] -- {} -- {:#06x}\n",
                 pc,
+                pc - 0x01_0000,
```
Why is this needed?

dezuman replied:
Oh, figured it would be helpful for tracing larger programs to calculate the byte offset from program start and show as decimal.

dezuman committed:
Correct minor formality

It doesn't actually matter since we're doing thumbs mode only (-> lowest bit always 1), but form is important.
